### ISSUE TITLE
HOC returned component props can not differ from HOC generic props

### ISSUE THREAD

**Issue by hpohlmeyer** [microsoft/TypeScript#28938]
**TypeScript Version:**  3.3.0-dev.20181208

**Search Terms:**
- HOC
- react
- higher order component

**Code**

```tsx
import * as React from 'react';

export interface HOCProps {
  foo: number;
}

/** Remove props, that have been prefilled by the HOC */
type WithoutPrefilled<T extends HOCProps> = Pick<T, Exclude<keyof T, 'foo'>>;

function withFoo<P extends HOCProps>(WrappedComponent: React.ComponentType<P>) {
  return class SomeHOC extends React.Component<WithoutPrefilled<P>> {
    public render(): JSX.Element {
      return <WrappedComponent {...this.props} foo={0} />;
    }
  };
}
```

**Expected behavior:**
No error, like with every version below 3.2.0.

**Actual behavior:**
Throws an error highlighting the WrappedComponent in the render method.
```
[ts]
Type 'Readonly<{ children?: ReactNode; }> & Readonly<Pick<P, Exclude<keyof P, "foo">>> & { foo: number; }' is not assignable to type 'IntrinsicAttributes & P & { children?: ReactNode; }'.
  Type 'Readonly<{ children?: ReactNode; }> & Readonly<Pick<P, Exclude<keyof P, "foo">>> & { foo: number; }' is not assignable to type 'P'. [2322]
```

**Additional Information**

This is pretty much the same sample example used in https://github.com/Microsoft/TypeScript/issues/28720, but with the difference, that the props of the returned component differ from the generic.

Basically the HOC prefills the `foo` property for the `WrappedComponent`. Since the spreaded props are overriden by `foo`, I don’t want `foo` to be a valid property for the HOC. This does not seem to be possible anymore.

**Playground Link:** 

**Related Issues:** 
https://github.com/Microsoft/TypeScript/issues/28720

**Comment by flsilva** [comment]
I believe I'm having the exact same problem on TypeScript 3.2.2, and I'd like to share my use case to show another way this is affecting HOCs in React. In my case I have a HOC that grabs `WrappedComponent`'s `children` to render it in a different way, forwarding all props but children to `WrappedComponent`, resulting in the exact same error @hpohlmeyer mentioned.

Here's a simplified (React Native) code:

```ts
import * as React from 'react';
import { ViewProps } from 'react-native';

export const withXBehavior = <P extends ViewProps>(
  WrappedComponent: React.ComponentType<P>,
): React.ComponentType<P> => (props) => {
  const { children, ...otherProps } = props;

  return (
    <WrappedComponent {...otherProps}> /* ERROR HERE */
      <React.Fragment>
        <View>
          <Text>example of HOC stuff here</Text>
        </View>
        {children}
      </React.Fragment>
    </WrappedComponent>
  );
};
```

Error:
`Type '{ children: Element; }' is not assignable to type 'P'. [2322]`

This was working fine on 3.0.1 and is now giving me this error on 3.2.2 (haven't tested on 3.1).

In my real world case I need `<P extends ViewProps>` to have the ability to use `onLayout`:

`<WrappedComponent {...otherProps} onLayout={this.myFunction} />`

If I forward all HOC props to WrappedComponent the error is gone, but that would prevent HOCs to change behavior of WrappedComponents.

```ts
import * as React from 'react';
import { Text, View, ViewProps } from 'react-native';

export const withXBehavior = <P extends ViewProps>(
  WrappedComponent: React.ComponentType<P>,
): React.ComponentType<P> => (props) => {
  const { children } = props;

  return (
    <WrappedComponent {...props}>
      <React.Fragment>
        <View>
          <Text>example of HOC stuff here</Text>
        </View>
        {children}
      </React.Fragment>
    </WrappedComponent>
  );
};
```

No errors in the above code.

**Comment by SpiderSkippy** [comment]
I'm seeing the same thing as the original poster, but from version 3.1.6 to 3.2.2

**Comment by weswigham** [comment]
@ahejlsberg The change is that we no longer erase generics in JSX (so we actually check these calls now), and roughly that `Pick<P, Exclude<keyof P, "foo">> & { foo: P["foo"] }` doesn't recombine to (or get recognized as assignable to) `P`. It's an unfortunate interaction with generic rest/spread, and the error we output is bad, too.

**Comment by hpohlmeyer** [comment]
Just for the sake of completeness: In some cases we do not remove the prop from the returned component, but make it optional instead. This also fails:

```tsx
import * as React from 'react';

export interface HOCProps {
  foo: number;
}

/** From T make every property K optional */
type Partialize<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>> & Partial<Pick<T, K>>;

/** Remove props, that have been prefilled by the HOC */
type OptionalPrefilled<T extends HOCProps> = Partialize<T, keyof HOCProps>;

export default function<P extends HOCProps>(WrappedComponent: React.ComponentType<P>) {
  return class SomeHOC extends React.Component<OptionalPrefilled<P>> {
    public render(): JSX.Element {
      return <WrappedComponent foo={0} {...this.props} />;
    }
  };
}
```

Error:

```
[ts]
Type 'Readonly<{ children?: ReactNode; }> & Readonly<Partialize<P, "foo">> & { foo: number; }' is not assignable to type 'IntrinsicAttributes & P & { children?: ReactNode; }'.
  Type 'Readonly<{ children?: ReactNode; }> & Readonly<Partialize<P, "foo">> & { foo: number; }' is not assignable to type 'P'. [2322]
```

So basically it’s the same. I just wanted to provide another example where TypeScript does not recognize that the type is assignable to T. This could be harder to solve then the original one though.

**Comment by sveyret** [comment]
It looks like the same problem as in #28884. It fails even if no spread rest.

**Comment by ChristianIvicevic** [comment]
Starting with 3.2 the [behaviour of the spread operator has changed](https://blogs.msdn.microsoft.com/typescript/2018/11/29/announcing-typescript-3-2/#object-spread-on-generic-types). Apparently the type of `props` gets erased as a negative side effect, but you can work around that by casting it back to `P` using `{...props as P}` when spreading back into the wrapped component.

**Comment by OliverJAsh** [comment]
> Apparently the type of `props` gets erased as a negative side effect, but you can work around that by casting it back to `P` using `{...props as P}` when spreading back into the wrapped component.

Is there a bug somewhere where we can track this specific problem?

**Comment by OliverJAsh** [comment]
> @ahejlsberg The change is that we no longer erase generics in JSX (so we actually check these calls now), and roughly that `Pick<P, Exclude<keyof P, "foo">> & { foo: P["foo"] }` doesn't recombine to (or get recognized as assignable to) `P`. It's an unfortunate interaction with generic rest/spread, and the error we output is bad, too.

@weswigham Is this the same as https://github.com/Microsoft/TypeScript/issues/28748?

**Comment by aprilandjan** [comment]
Had to use `<WrappedComponent {...this.props as any} inject={injected}/>` to temporarily bypass the HOC props merge.

**Comment by lvkins** [comment]
I think that this issue can be closed, since the implicit types of the spread operator were changed.

This is a fully working example you can use with TS 3.2+

```
type ShadowInjected<T1, T2> = Omit<T1, keyof T2>;

interface HOCProps {
    foo : number;
}

export const withSomething = <T extends HOCProps>(WrappedComponent: React.ComponentType<T>): React.FC<ShadowInjected<T, HOCProps>> => {
    return function WithSomething(props: ShadowInjected<T, HOCProps>) {
        // Do you HOC work here
        return (<WrappedComponent foo={1337} {...props as T} />);
    };
};
```

Of course, if you need to shadow only one property, you don't need the `ShadowInjected` type.

**Comment by gitschwifty** [comment]
I'm not sure if this is specific to my environment @lvkins , but I was getting 'keyof T2 does not satisfy the constraint keyof T1' which dropped to 'type string is not assignable to type keyof T1'. Fixed by specifying T1 as an extension of T2:

`type ShadowInjected<T1 extends T2, T2> = Omit<T1, keyof T2>`

Let me know if this is incorrect, am on 3.8.2, been using typescript for a while but sometimes I've troubles grokking the complexities of typing. Thanks!

**Comment by iamandrewluca** [comment]
@lvkins the idea here is to force `WrappedComponent` to extend your HOC injected props

**Comment by D34THWINGS** [comment]
I'm kind of reviving this thread but casting is not a proper solution, it's highly unsafe. Here's a sample:
```
import React, { ComponentType } from 'react'

export type BaseProps = {
  value: string
  name: string
}

export type InjectedProps = {
  onChange: () => {}
}

export const withOnChange = <TProps extends BaseProps & InjectedProps>(
  Component: ComponentType<TProps>
) => {
  return ({ name, ...props }: Omit<TProps, keyof InjectedProps>) => {
    return <Component {...(props as TProps)} />
  }
}
```
This sample is totally valid even `Component` does not receive either `onChange` and `name` which are **required** props. We ship a massive bug due to this in our product today because of this. This really needs to be fixed, casting is such a bad idea and I feel like I have to use it too many times. If someone has a better workaround solution, I would be happy about it

**Comment by satansdeer** [comment]
I've tried to do this to avoid type assertion:

```tsx
type InjectedProps = {
  foo: string
}

type PropsWithoutInjected<TBaseProps> = Omit<
  TBaseProps,
  keyof InjectedProps
>

export function withFoo<TProps>(
  WrappedComponent: React.ComponentType<
    PropsWithoutInjected<TProps> & InjectedProps
  >
) {
  return (props: PropsWithoutInjected<TProps>) => {
    return <WrappedComponent {...props} foo="bar" />
  }
}
```

This way the wrapped component can't receive the injected props and TypeScript does not complain.

Here is a [TypeScript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wFgAoUSWRYmAEQHkBZObXAo9GAWgBNcZchRgBPMEjgBJAHYArJBiR8ACjjABnOAF44Abwpw2ECAC44GmFGAyA5hQC+FEeMlqImgOrAYACwgArjCyCkp8ADwAKgBCKBpI7poAfDpwjCA+4YZwMXEJ6hoANNkA1kiiEJjS8oowyokaFEnO5EgAHtRYATIYwBAycADuPr4AYiZRDUkAFNmeUChgEnwAwriQMkgyMObI3AB0a+D9WzCRrlnkRkYN3n6BwTVhkwUpAGTVoXWqBdnN5ABKfTZIgwAJQAbTMAFcy3EYPEK1ZQvDwaJJA7QpAxXa6g8EDcLzRbLI4bU76faU6GohzGCDafAAIxQUHwcAA9P8jE5yDyXBI4ABRNoocAAGyQpJO2waqWx1zp5ks1js2SMzKgSqsNnsvJaaH6liFIvFkvW0p2tAOoxW4WForAEqlmxlr1S0z0mBMDgxKXCYCS4wgcGAWk93vC7IDAG4KAaZEb7abnac7mMTKlhn4g3aTY6zccXTAptMk-mU9sAS09hgmMx9kQZHwkFBprmHU7zUW00G4BqGXVLPhOYU4DIAmKxVXyEA) with an example of how this HOC can be used.

**Comment by gabriellsh** [comment]
@satansdeer but you're not getting the rest props, it should be `return ({somekey, ...props}) =>...`.

**Comment by satansdeer** [comment]
@gabriellsh not sure why do you need to get the `rest` as you are overriding the `foo` prop in my example. React will accept the last value of the prop  that was passed, here is a [codesandbox](https://codesandbox.io/s/fast-water-dthsv?file=/src/App.js).

**Comment by satansdeer** [comment]
Ok here is another take from @dex157:

```tsx
export interface WithFooProps {
  foo: number;
}

export function withFoo<ComponentProps>(WrappedComponent: React.ComponentType<ComponentProps & WithFooProps>) {
  return class FooHOC extends React.Component<ComponentProps> {
    public render() {
      return <WrappedComponent {...this.props} foo={0} />;
    }
  };
}
```

This way we don't even have to use `Omit` and type inference works correctly:

```tsx
interface ButtonProps {
  title: string;
}

const Button: React.ComponentType<ButtonProps> = withFoo(({title, foo}) => {
  return <button id={String(foo)}>{title}</button>
})


const App = () => {
  return <main><Button title="click me" /></main>
}
```

Here is a [TypeScript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwCwAUI1gB6SxzAB2MWUBeLHADqwGAAsAYhAgAFYmHQBvRnDgEZALjhcAriABGfegwC+jZm2iFdXfMAhc4AdzFSZAHgDCJSFyw88hCKAHwAFMJQyGBgWAAmPuCOATDa2HgwAHSJfikAKgCesd6+yYEK6ABkIm7SchUhAJRwKgxqlDC6UE5hYBXaOWUwQYrNALwhLapqcB1dTh6R0bEJpf48LZlbfcGopuoyY0oADPsA9CEmauZmJoxnZ3AAoizI4AA2WBYM3Lz8gnAAEK6GAwRwjZTTGBiT7aVAwKDcADmJhujFwjnhQJBYK4aSoWUG6xghWKwNB4IacDGLlqMjCYSU0JgnwANAcIKZxpNWu0sJ1unAPAYcY5OHEjgBlBHIsIaCCNUwhJkwrCmDxnEUUrghRhc77ozHwACCMWpcDC3KmbVm-PmQpAyG4IQ85NxcGZnzGACJcO9gLgANZwEBYb1wC4ax3OvVAA) to try it out.

**Comment by coalaroot** [comment]
Basically the difference in @satansdeer answer is to use type (which is not stated there explicitly) instead of interfaces

This will **not work**
```
import * as React from 'react';

export interface WithFooProps {
  foo: number;
}

interface ButtonProps extends WithFooProps {
  title: string;
}

export function withFoo<ComponentProps>(WrappedComponent: React.ComponentType<ComponentProps & WithFooProps>) {
  return (props: ComponentProps) => {
    return <WrappedComponent {...props} foo={0} />;
  }
};

const RealButton: React.FC<ButtonProps> = ({title, foo}) => {
  return <button id={String(foo)}>{title}</button>
} 

const Button = withFoo(RealButton)


const App = () => {
  return <main><Button title="click me" /></main>
}
```

And this **will**

```
import * as React from 'react';

export interface WithFooProps {
  foo: number;
}

type ButtonProps= {
  title: string;
} & WithFooProps

export function withFoo<ComponentProps>(WrappedComponent: React.ComponentType<ComponentProps & WithFooProps>) {
  return (props: ComponentProps) => {
    return <WrappedComponent {...props} foo={0} />;
  }
};

const RealButton: React.FC<ButtonProps> = ({title, foo}) => {
  return <button id={String(foo)}>{title}</button>
} 

const Button = withFoo(RealButton)


const App = () => {
  return <main><Button title="click me" /></main>
}
```

The difference is in **ButtonProps**

**Comment by satansdeer** [comment]
@coalaroot great catch. Yes, using the `extends` syntax is not the same as having an intersection type.

**Comment by Omelyan** [comment]
> And this **will**

will... how about generics in FooProps? :))
```typescript
export interface WithFooProps<T> {
  foo: T;
}
```

**Comment by coalaroot** [comment]
Can do but why though?

**Comment by Omelyan** [comment]
@coalaroot I can't)

**Comment by coalaroot** [comment]
Sorry man, don't have the time to figure it out :(

**Comment by svidskiy** [comment]
Any updates?

**Comment by lifeiscontent** [comment]
@hayata-suenaga weird thing is, this works:

```ts
type Foo = {
  foo: string;
};

function addFoo<P extends Foo>() {
  return (argWithoutFoo: Omit<P, keyof Foo>) => {
    const argWithFoo: Omit<P, keyof Foo> & Pick<P, keyof Foo> = { ...argWithoutFoo, foo: "foo" };
  };
}
```

**Comment by keropodium** [comment]
In my case I had to go from this:
``` tsx
export function withBaseProviders<P extends {}>(
  WrappedComponent: React.ComponentType<P>,
) {
  const WithBaseProviders = ({ a, b, c, d, e, ...props }: P & BaseProvidersProps) => (
    <BaseProviders a={a} b={b} c={c} d={d} e={e}>
      <WrappedComponent {...props} />
    </BaseProviders>
  );
  WithBaseProviders.displayName = `BaseProviders(${
    WrappedComponent.displayName || WrappedComponent.name || "Component"
  })`;
  return WithBaseProviders;
}
```
That raised:
```
Type 'Omit<P & BaseProvidersProps, "a" | "b" | "c" | "d" | "e">' is not assignable to type 'IntrinsicAttributes & P'.
  Type 'Omit<P & BaseProvidersProps, "a" | "b" | "c" | "d" | "e">' is not assignable to type 'P'.
    'Omit<P & BaseProvidersProps, "a" | "b" | "c" | "d" | "e">' is assignable to the constraint of type 'P', but 'P' could be instantiated with a different subtype of constraint '{}'.ts(2322)
```

To extend the generic type and then explicitly omit on the wrapped:
``` tsx
export function withBaseProviders<P extends BaseProvidersProps>(
  WrappedComponent: React.ComponentType<Omit<P, keyof BaseProvidersProps>>,
) {
  const WithBaseProviders = ({ a, b, c, d, e, ...props }: P) => (
    <BaseProviders a={a} b={b} c={c} d={d} e={e}>
      <WrappedComponent {...props} />
    </BaseProviders>
  );
  WithBaseProviders.displayName = `BaseProviders(${
    WrappedComponent.displayName || WrappedComponent.name || "Component"
  })`;
  return WithBaseProviders;
}
```

**Comment by RyanCavanaugh** [comment]
This error is correct. `P extends HOCProps` does not mean that `P` is exactly `HOCProps`; it can require a narrower `foo` type. For example, `P` could be `{ foo: 1; label: string }`. The HOC supplies `foo={0}`, so the props assembled from `WithoutPrefilled<P>` and `{ foo: number }` are not safe to pass to every possible `React.ComponentType<P>`.

This is the distinction described in [Why can't I narrow or assign to a generic type parameter the way I can with its constraint?](https://github.com/microsoft/TypeScript/wiki/FAQ#why-cant-i-narrow-or-assign-to-a-generic-type-parameter-the-way-i-can-with-its-constraint). TypeScript 3.2 correctly began checking the generic JSX call instead of erasing its type parameter.
